### ISSUE TITLE
Prerequisite libraries

### ISSUE THREAD

**Issue by 3246251196** [AmigaLabs/webkitty#4]
We need the following libraries to be available in the SDK, compiled for clib2

Most of them are available at OS4Depot, but we need to check if they are usable from clib2 or we need to compile them ourselves
- [x] jpeg 
- [x] png
- [x] xslt 
- [x] sqlite
- [x] xml2
- [x] libICU

We can also request from others to work on the port of these libraries, for example there is a port of libICU since 2013 and I contacted the person who did it if he is available to do a new one for us

- [ ] Log the process at repo wiki pages

**Comment by 3246251196** [comment]
Update on libICU:

With:
`../configure CC="ppc-amigaos-gcc -mcrt=clib2" CXX="ppc-amigaos-g++ -mcrt=clib2" CPPFLAGS="-mcrt=clib2 -athread=native" --disable-tests --disable-samples --host=ppc-amigaos --disable-threads --disable-shared --enable-static --disable-dyload --enable-release --disable-extras --with-data-packaging=files --with-cross-build=/home/rjd/projects/webkitty_proj/icu/icu4c/source/buildLinux/ THREADSCXXFLAGS=-athread=native`

Able to get:
```
./tools/ctestfw/libicutest.a
./stubdata/libicudata.a
./lib/libicuio.a
./lib/libicuuc.a
./lib/libicutu.a
./lib/libicui18n.a
```

Waiting for confirmation that this is expected, or that something is missing.

For now, I will work on other libraries.

==
Update
==
https://github.com/3246251196/icu
in icu4c there is a script named "rjdMake" : that is how I built the libraries. I used: https://github.com/3246251196/adtools/tree/adtools_be (which is adtools + call_once fix + SDK 54.16 + HEAD@clib2 afxgroup) as the compiler.

As I say, I will now continue to work on other libraries.

**Comment by walkero-gr** [comment]
I tried today the libICU from the Odyssey repository and I get the following error.

```
Failed to find all ICU components: Found unsuitable version "49.1.2", but
  required is at least "61.2" (found
```

This means we have to port the latest version, so to fulfill this requirement.

**Comment by 3246251196** [comment]
### ICU
At the moment, I am building this library (well, it is a collection of libraries) by cross-compilation and then running the pkgdata/genccode on the AmigaOne due to the issue with BUILD/HOST endiannes. I have added a number of hacks into the "rjdMake" file.

The latest version of this library is: https://github.com/3246251196/icu/blob/main/icu4c/source/libicu.lha.

Testing it is not just a simple matter of running the ICU tests like: ./RunTest1.exe and waiting for a response - those programs need the correct paths to the data files and this and that. Of course, we should test it. But, right now, I think I will assume it all works and move on to xml2 - which is not idea, I know.

[[ TODO: Consider raising a PR with official ICU repo to fix needed byte swapping for BUILD/HOST endiannes lack of support ]]

**Comment by 3246251196** [comment]
### LIBXML2
I have built the latest LIBXML2 using the same stuff as for ICU. LIBXML iconv disabled, since we link with our recently build ICU library. Added "rjdMake" to that repository too: https://github.com/3246251196/libxml2 with the latest version of the library: https://github.com/3246251196/libxml2/blob/master/libxml2.lha.

**Comment by 3246251196** [comment]
### SQLITE3

See: https://github.com/3246251196/sqlite/blob/master/libsqlite3.lha

**Comment by 3246251196** [comment]
### XSLT

Can we use the recent library on Os4Depot: http://os4depot.net/?function=showfile&file=development/library/libxslt.lha (built with GCC 4.2.4 and SDK5334) ; includes clib2)

**Comment by 3246251196** [comment]
### LIBPNG

See: https://github.com/3246251196/libpng/blob/libpng16/libpng.lha

**Comment by 3246251196** [comment]
### LIBJPEG

Can we use the recent library on Os4Depot: http://os4depot.net/?function=showcontent&file=development/library/graphics/libjpeg.lha (includes clib2)

**Comment by 3246251196** [comment]
@walkero-gr 
I do need to go back and build the newlib versions of these. finally, creating an uploaded DEV package to OS4DEPOT. Also, I will invoke the built tests per library where I can: this I also have to do.

But, in the meantime, can we test all of these above with a CLIB2 webkit?

TODO: Correct pkgconfig .pc files since they will take my temporary installation prefix paths - need to change this to real amiga paths or install to the actual SDK

**Comment by 3246251196** [comment]
Update: NEWLIB and CLIB2 versions of all libraries should now be available:

jpeg (ALREADY AVAILABLE ON OS4DEPOT within 3 years)
png (https://github.com/3246251196/libpng/blob/libpng16/libpng.lha)
xslt (ALREADY AVAILABLE ON OS4DEPOT within 3 years)
sqlite (https://github.com/3246251196/sqlite/blob/master/libsqlite3.lha)
xml2 (https://github.com/3246251196/libxml2/blob/master/libxml2.lha)
libICU (https://github.com/3246251196/icu/blob/main/icu4c/source/libicu.lha)

The next stage is to test them out.

@walkero-gr

**Comment by kas1e** [comment]
@3246251196 
Generally speaking, you need to test only sqlite and libICU (as they can act); other ones, such as jpeg, png, xslt, and xml2, require no amigaOS4-specific changes, and pure configuring them to build for OS4 works (that's how all the versions of them build on os4depot, and that's how I build them for Odyssey as well).

The only issue is sqlite (as a pure build, it will not work as expected and will require amiga-specific changes). For Odyssey, I use http://aminet.net/package/biz/dbase/sqlite-3.34.0-amiga. Lately, I just use AmigaOS' ifdefs from this version to add them on top of the latest sqlite.

And libICU can bring some problems too; if I remember right, it required a little bit of change as well.

**Comment by 3246251196** [comment]
@walkero-gr 
Will need to help in this QR section. I am not able to test ICU without some additional help.

==
To elaborate: I think the issue comes down to the data files being dumped for little-endian machines. When I build ICU, I follow the guide as linked: https://unicode-org.github.io/icu/userguide/icu4c/build.html#how-to-cross-compile-icu. This build the Linux stuff fine, of course. Then comes the Amiga. Everything builds fine before the binary "pkgdata" is invoked (this one is the pkgdata for the Linux build - that is how the cross compilation works). "pkgdata" says: different endianess is not supported. As already said above, to get around this I manually run that step in the build process on my AmigaOne machine with all the inputs etc and finally arrive at the library file libicudata.a. Then, I copy this file over to the Linux machine, hack the makefile and get the full build complete.

Now, we want to run the ICU tests to ensure completeness. The issue, as I have remarked above, is that the tests are not just self contained simple stand-alone executable files. They rely on data in directories higher than the test directory itself. It is kind of a spaghetti dependency so it is not trivial. I have been working on the test named "iotest" on the AmigaOne. I get the error "U_INVALID_FORMAT_ERROR" - but since I am not a beta tester I have no access to a debugger and probably will have to resort to printf debugging whilst debugging the Linux build to know where to insert printfs. This is not ideal. The tests are essentially set up to expect to be run on the machine which has just built the library - this makes it quite an ordeal to recreate the same folder structures on the AmigaOne.

I also do not know if the binary named "genrb" should be outputting binary data in a format that can be consumed by a big-endian machine. If this is the case, though, I wonder why no error is generated when invoking "genrb" from linux via the Amiga building context.

Basically, I could use some help here. Regardless, I will plod along, but without really understanding ICU more and without the experience of doing this previously, it may take a while to test ICU.

===

Just a note: I have since read that you can distinguish the endianess in datas by the folder name ending with an "l" or a "b". Such as: icudt72b vs icudt72l.

**Comment by 3246251196** [comment]
https://github.com/3246251196/libjpeg/releases/tag/AmigaOS4_9e
https://github.com/3246251196/libpng/releases/tag/AmigaOS4_1_6_39
https://github.com/3246251196/libxslt/releases/tag/AmigaOS4_1_1_37
https://github.com/3246251196/sqlite/releases/tag/AmigaOS4_3_34
https://github.com/3246251196/libxml2/releases/tag/AmigaOS4_2_10_3
https://github.com/3246251196/icu/releases/tag/AmigaOS4_72_1

This is what I have so far. I believe they each have CLIB2 and NEWLIB versions, built with:
GCC 11.3, binutils 2.23.2, Afxgroup CLIB2 (no way to find out which version was used, but it was late last year).

I would really need help with actually testing whether or not these libraries work. A lot of these tests just assume that they are running on the HOST machine which is obviously not the case for us. Our host machine is the Amiga. Quite some work would be necessary.

Anyway, this is where I am so far.

@walkero-gr

**Comment by afxgroup** [comment]
I'm trying to recompile all of  this libraries with shared clib2 and i've created an ubuntu package for all of them.
But while almost of them are easy to compile, ICU is a pain in the.. However compiling right now. Then I'll try to test it

**Comment by 3246251196** [comment]
I guess we got all of these built. I wonder if you ran the tests for them or just built them?
